game/scripts/vscripts/units/shielder.lua

```diff
@@ -0,0 +1,212 @@
+
+--[[ Taken from bb template
+if ShielderAI == nil then
+  DebugPrint ( 'creating new BossAI object' )
+  ShielderAI = class({})
+
+  Debug.EnabledModules['boss:*'] = false
+end]]
+
+LinkLuaModifier( "modifier_boss_phase_controller", "modifiers/modifier_boss_phase_controller", LUA_MODIFIER_MOTION_NONE )
+
+function Spawn (entityKeyValues)
+  thisEntity:FindAbilityByName("boss_shielder_shield")
+
+  thisEntity:SetContextThink( "ShielderThink", partial(ShielderThink, thisEntity) , 1)
+  print("Starting AI for " .. thisEntity:GetUnitName() .. " " .. thisEntity:GetEntityIndex())
+
+  ABILITY_shield = thisEntity:FindAbilityByName("boss_shielder_shield")
+
+  local phaseController = thisEntity:AddNewModifier(thisEntity, ABILITY_shield, "modifier_boss_phase_controller", {})
+  phaseController:SetPhases({ 66, 33 })
+  phaseController:SetAbilities({
+    "boss_shielder_shield"
+  })
+end
+
+--[[
+ShielderAI.IDLE = 1
+ShielderAI.AGRO = 2
+ShielderAI.LEASHING = 3
+ShielderAI.DEAD = 4
+]]
+
+function Create (unit, options)
+  options = {}
+  options.tier = 1
+
+  local state = {
+    handle = unit,
+    origin = unit:GetAbsOrigin(),
+    leash = 1500,
+    agroDamage = 100,
+    tier = 1,
+    currentDamage = 0,
+    state = 1,
+    customAgro = true,
+
+    deathEvent = Event()
+  }
+
+  unit:OnHurt(function (keys)
+    HurtHandler(state, keys)
+  end)
+  unit:OnDeath(function (keys)
+    DeathHandler(state, keys)
+  end)
+
+  unit:SetIdleAcquire(false)
+  unit:SetAcquisitionRange(0)
+
+  return {
+    onDeath = state.deathEvent.listen
+  }
+end
+
+function HurtHandler (state, keys)
+  if state.state == 1 then
+    DebugPrint('Checking boss agro...')
+    DebugPrintTable(keys)
+
+    state.currentDamage = state.currentDamage + keys.damage
+
+    if state.currentDamage > state.agroDamage then
+      Agro(state, EntIndexToHScript(keys.entindex_attacker))
+      state.currentDamage = 0
+    end
+  elseif state.state == 2 then
+  end
+end
+
+function GiveItemToWholeTeam (item, teamId)
+  for playerId = 0,19 do
+    if PlayerResource:GetTeam(playerId) == teamId and PlayerResource:GetPlayer(playerId) ~= nil then
+      local player = PlayerResource:GetPlayer(playerId)
+      local hero = player:GetAssignedHero()
+
+      hero:AddItemByName(item)
+    end
+  end
+end
+
+function DeathHandler (state, keys)
+  DebugPrint('Handling death of boss ' .. state.tier)
+  state.state = 4
+
+  state.handle = nil
+  local killer = EntIndexToHScript(keys.entindex_attacker)
+  local teamId = killer:GetTeam()
+
+  state.deathEvent.broadcast(keys)
+
+  if teamId == 2 then
+    team = 'good'
+  elseif teamId == 3 then
+    team = 'bad'
+  else
+    return
+  end
+
+  GiveItemToWholeTeam("item_upgrade_core", teamId)
```

Comment:
The existing death handler should already handle giving out the upgrade cores. This code can be removed.